### PR TITLE
Preventing stack overflow exceptions via limiting the depth of the parser rules

### PR DESCRIPTION
**Pull request by bbakerman**
This puts in a limit on how deep grammar rules can get.  

As ANLTR is a **recursive-descent parser** it will build up a call stack as it moves along.  This stack is limited and can blow up with just the wrong input.  Even if the max tokens checking is in place, it can still blow the JVM stack before the max tokens are reached.

Right now it's 500 deep as presented here.  Is this the right value?  More?? Less ??  I changed it from 1000

### CODE REVIEW

File: src/test/groovy/graphql/parser/ParserTest.groovy [hunk 1]
```diff
@@ -40,31 +39,15 @@ import graphql.language.TypeName
...
-        ParserOptions.setDefaultParserOptions(defaultOptions)
-        ParserOptions.setDefaultOperationParserOptions(defaultOperationOptions)
-        ParserOptions.setDefaultSdlParserOptions(defaultSdlOptions)
-    }
-
-    void cleanup() {
-        ParserOptions.setDefaultParserOptions(defaultOptions)
-        ParserOptions.setDefaultOperationParserOptions(defaultOperationOptions)
-        ParserOptions.setDefaultSdlParserOptions(defaultSdlOptions)
-    }
```
Comment (bbakerman) [inline src/test/groovy/graphql/parser/ParserTest.groovy:66]: Moved to the Stress test class along with other stress tests

File: src/test/groovy/graphql/parser/ParserTest.groovy [hunk 2]
```diff
@@ -404,7 +383,8 @@ class ParserTest extends Specification {
                 .build()
 
         when:
-        def document = new Parser().parseDocument(input, parserOptionsWithoutCaptureLineComments)
+        def parserEnvironment = newParserEnvironment().document(input).parserOptions(parserOptionsWithoutCaptureLineComments).build()
+        def document = new Parser().parseDocument(parserEnvironment)
```
Comment (bbakerman) [inline src/test/groovy/graphql/parser/ParserTest.groovy:387]: removed deprecated methods

File: src/test/groovy/graphql/parser/ParserTest.groovy [hunk 3]
```diff
@@ -1158,112 +1142,14 @@ triple3 : """edge cases \\""" "" " \\"" \\" edge cases"""
...
         options = ParserOptions.newParserOptions().captureSourceLocation(false).build()
-        document = new Parser().parseDocument("{ f }", options)
+        def parserEnvironment = newParserEnvironment().document("{ f }").parserOptions(options).build()
+        document = new Parser().parseDocument(parserEnvironment)
 
         then:
         !options.isCaptureSourceLocation()
         document.getSourceLocation() == SourceLocation.EMPTY
         document.getDefinitions()[0].getSourceLocation() == SourceLocation.EMPTY
     }
```
Comment (bbakerman) [inline src/test/groovy/graphql/parser/ParserTest.groovy:1153]: Moved the following the ParserStressTest

**Review by dondonz — APPROVED** [review]

**Review by temaEmelyan — APPROVED** [review]

**Review by russellyou — APPROVED** [review]

File: src/main/java/graphql/parser/exceptions/ParseCancelledTooDeepException.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package graphql.parser.exceptions;
+
+import graphql.Internal;
+import graphql.i18n.I18n;
+import graphql.language.SourceLocation;
+import graphql.parser.InvalidSyntaxException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+@Internal
```
Comment (andimarek) [inline src/main/java/graphql/parser/exceptions/ParseCancelledTooDeepException.java:10]: should this be public because the Parser throws it?

**Review by andimarek — APPROVED** [review]
see one comment about Internal exception ... otherwise LGTM

File: src/main/java/graphql/parser/exceptions/ParseCancelledTooDeepException.java (hunk 4 above)
Comment (bbakerman) [reply src/main/java/graphql/parser/exceptions/ParseCancelledTooDeepException.java:10]: well it has to be since its in another package (with the others)

```
exceptions
  InvalidUnicodeSyntaxException
  MoreTokensSyntaxException
  ParseCancelledException
  ParseCancelledTooDeepException
```
